### ISSUE TITLE
yahoo.symbol.details

### ISSUE THREAD

**Issue by anycorp** [mkfs/misc-finance#1]
I tried using this function but it doesn't work.  Wonder if there's an updated version re the same?  many thanks!

**Comment by mkfs** [comment]
Haven't used this code in years, but I'll take a look. Original project went in another direction once sentiment analysis proved to be as reliable as any social-media data source.

**Comment by anycorp** [comment]
many thanks!

**Comment by mkfs** [comment]
Looks like they've moved to react.js, so it's not using the same DOM as before. 
The keys/values from the lines
```
 keys <- xpathSApply(html, 
              "//table[@class='yfnc_datamodoutline1']//td[@class='yfnc_tablehead1']", xmlValue)
  vals <- xpathSApply(html, 
              "//table[@class='yfnc_datamodoutline1']//td[@class='yfnc_tabledata1']", xmlValue)
```
...are probably from the Statistics pane. The table and td class values are formatting descriptions rather than content desriptions, so the XPath will have to first find the div that contains them ("quote-summary" ?) and then select the first table object.

The HTML for this part of the table is as follows:
```
<div id="quote-summary" class="Bxz(bb) D(ib) Va(t) Mih(250px)--lgv2 W(100%) Mt(-6px) W(47.674%)--lgv2 Mend(40px)--lgv2" data-test="quote-summary-stats">
<div class="D(ib) W(1/2) Bxz(bb) Pend(10px) Va(t)" data-test="left-summary-table">
<table class="W(100%)">
<tbody>
<tr class="Bxz(bb) Bdbw(1px) Bdbs(s) Bdc($lightGray) H(36px) ">
<td class="C(black)">
<span>Previous Close</span>
</td>
```

**Comment by anycorp** [comment]
Apology, I am a beginner in R and don't know much about html either. I don't quite get what I need to do to change the code to extract the Sector/Industry info from the Yahoo Finance Summary page.  For example, for the path http://finance.yahoo.com/quote/1044.HK?p=1044.HK, 

I found the xpath for the Sector and Industry info to be below:
//*[@id="main-0-Quote-Proxy"]/section/div[2]/aside/div[4]/div/div/p[2]/strong[1]
//*[@id="main-0-Quote-Proxy"]/section/div[2]/aside/div[4]/div/div/p[2]/strong[2]

so I wonder how do i change the keys/val line to extract the sector and the industry info from that page please?

Thank you so much for your patience/help and apology for my ignorance in both R and html.

**Comment by mkfs** [comment]
OK, I'l play with the selectors when I get a chance, and see if I can get it to work. XML parsing always ends up being trial-and-error.

**Comment by mkfs** [comment]
Looks like the content is loaded dynamically - if you download 'https://finance.yahoo.com/quote/MSFT/key-statistics?p=MSFT' with curl, for example, you will see a Loading... placeholder where the statistics table should be.

This means that a Javascript interpreter must be present in order to fetch the statistics for parsing. That is well beyind the capability of R's htmlParse library (and Curl, for that matter). It might be feasible to use a console-based browser like w3m to load the page and dump it to STDOUT, but that's a lot more work than I am illing to put into this.

**Comment by anycorp** [comment]
the Sector/Industry info seems to be under the Profile tab ('https://finance.yahoo.com/quote/MSFT/profile?p=MSFT') not key-statistics.  does that make any difference?

**Comment by mkfs** [comment]
Yeah, that's still a JS-populated page. Curl just shows "Loading..." for the details. Seleneium is the way to go, but I don't know when I'll have time to make that extensive of a change.

**Comment by mkfs** [comment]
Took a look at this again tonight, now that I have some free time. Either Curl (and/or Rcurl) upgraded itself to handle React.js documents, or I did something seriously wrong with my previous parsing attempts.

It looks like the data is (now?) accessible:
```
> msft.url <- 'https://finance.yahoo.com/quote/MSFT/key-statistics?p=MSFT'
> msft.html <- getURLContent(msft.url, followlocation = TRUE, binary=FALSE)
> msft.xml <- htmlParse(msft.html)
> xpathSApply(msft.xml, "//td", xmlValue)
  [1] ""                                  "Search"                           
  [3] "Market Cap (intraday) 5"           "572.58B"                          
  [5] "Enterprise Value 3"                "N/A"                              
  [7] "Trailing P/E "                     "27.43"                            
  [9] "Forward P/E 1"                     "20.59"                            

```
Not ony is the information available via getURLContent, it looks pretty simple to parse as well: just find the index of relevant strings and extract index+1 as the value. Will try to get a patch out over the weekend.

**Comment by anycorp** [comment]
Thanks!

**Comment by mkfs** [comment]
Here are the xpaths:

```
xpathSApply(msft.stats.xml, "//div[@id='Main']//td", xmlValue)
xpathSApply(msft.profile.xml, "//div[@class='asset-profile-container']//*[self::span or self::strong]", xmlValue)
```                           
I'll update the functions to use these (and to hit both URLs) and test on a few stock symbols before I push a fix.

**Comment by mkfs** [comment]
Should be fixed in bc0e00543f45e2a9fbc1cbe5f1c1316f6a7fd4b2 .

**Comment by mkfs** [comment]
Considering this closed. Re-open if the above patch does not fix parsing of current (Q3 2017) yahoo finance pages.
